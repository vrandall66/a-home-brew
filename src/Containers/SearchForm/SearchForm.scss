* {
  text-align: center;
}

.SearchForm {
  align-items: center;
  display: flex;
  flex-direction: column;
  height: 23vh;
  width: 100%;
}

.SearchForm__input--search-term,
.SearchForm__button--search {
  font-size: 100%;
}

.SearchForm__div--container {
  width: 50%;
}