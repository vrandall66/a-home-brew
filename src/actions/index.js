export const setBeers = beers => ({
  type: 'SET_BEERS',
  beers
});

export const addProperties = beers => ({
  type: 'ADD_PROPERTIES',
  beers
});

export const setLoading = bool => ({
  type: 'IS_LOADING',
  bool
});

export const setError = error => ({
  type: 'HAS_ERROR',
  error
});

export const toggleBookmark = beer => ({
  type: 'TOGGLE_BOOKMARK',
  beer
});

export const setBookmark = beer => ({
  type: 'SET_BOOKMARK',
  beer
});

export const removeBookmark = beer => ({
  type: 'REMOVE_BOOKMARK',
  beer
});

export const addCurrentlyBrewing = beer => ({
  type: 'ADD_CURRENTLY_BREWING',
  beer
});

export const resetFavorites = () => ({
  type: 'RESET_FAVORITES'
});
