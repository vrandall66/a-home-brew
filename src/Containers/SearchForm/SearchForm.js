import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import { setSearchResults } from '../../actions';
import { bindActionCreators } from 'redux';
import { PropTypes } from 'prop-types';
import './SearchForm.scss';

export class SearchForm extends Component {
  constructor() {
    super();
    this.state = {
      searchTerm: '',
      hops: '',
      malt: '',
      yeast: ''
    };
  }

  getHopsOptions = () => {
    const { beers } = this.props;
    return beers.reduce((filtered, beer) => {
      beer.ingredients.hops.forEach(i => {
        if (!filtered.includes(i.name)) {
          filtered.push(i.name);
        }
      });
      return filtered.sort();
    }, []);
  };

  getMaltOptions = () => {
    const { beers } = this.props;
    return beers.reduce((filtered, beer) => {
      beer.ingredients.malt.forEach(i => {
        if (!filtered.includes(i.name)) {
          filtered.push(i.name);
        }
      });
      return filtered.sort();
    }, []);
  };

  getYeastOptions = () => {
    const { beers } = this.props;
    return beers.reduce((filtered, beer) => {
      if (!filtered.includes(beer.ingredients.yeast)) {
        filtered.push(beer.ingredients.yeast);
      }
      return filtered.sort();
    }, []);
  };

  searchByHopsMaltOrYeast = async () => {
    const { setSearchResults } = this.props;
    const { hops, malt, yeast } = this.state;
    const baseUrl = 'https://api.punkapi.com/v2/beers?';
    const hopsQuery = hops ? `hops=${hops}&` : '';
    const maltQuery = malt ? `malt=${malt}&` : '';
    const yeastQuery = yeast ? `yeast=${yeast}&` : '';
    const response = await fetch(
      `${baseUrl}${hopsQuery}${maltQuery}${yeastQuery}`
    );
    if (!response.ok) {
      throw new Error(
        'Could not retrieve beers, please explore with us later.'
      );
    }
    const data = await response.json();
    return await setSearchResults(data);
  };

  searchBySearchTerm = async () => {
    const { setSearchResults } = this.props;
    const { searchTerm } = this.state;
    const baseUrl = 'https://api.punkapi.com/v2/beers?';
    const termQuery = searchTerm ? `beer_name=${searchTerm}` : '';
    const response = await fetch(`${baseUrl}${termQuery}`);
    if (!response.ok) {
      throw new Error(
        'Could not retrieve beers, please explore with us later.'
      );
    }
    const data = await response.json();
    return await setSearchResults(data);
  };

  handleChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  handleSubmit = e => {
    e.preventDefault();
    this.handleClick();
  };

  handleClick = () => {
    const { history } = this.props;
    const { hops, malt, yeast, searchTerm } = this.state;
    if (hops || malt || yeast) {
      this.searchByHopsMaltOrYeast();
    }
    if (searchTerm) {
      this.searchBySearchTerm();
    }
    this.clearInputs();
    return history.push('/search');
  };

  clearInputs = () => {
    const stateKeys = Object.keys(this.state);
    stateKeys.forEach(stateKey =>
      this.setState({
        [stateKey]: ''
      })
    );
  };

  render() {
    return (
      <>
        <form className='SearchForm' onSubmit={e => this.handleSubmit(e)}>
          <div className='SearchForm__div--container'>
            <input
              name='searchTerm'
              value={this.state.searchTerm}
              onChange={this.handleChange}
              placeholder='Search by name or description'
              className='SearchForm__input--search-term'
            ></input>
            <section className='SearchForm__section--select-container'>
              <section>
                <label htmlFor='hops' className='SearchForm__label'>
                  Hops:{' '}
                </label>
                <select
                  id='hops'
                  onChange={e => this.setState({ hops: e.target.value })}
                  value={this.state.hops ? this.state.hops : 'All'}
                  className='SearchForm__select'
                >
                  <option value='' id='all-hops'>
                    All
                  </option>
                  {this.getHopsOptions().map((hops, index) => {
                    return (
                      <option value={hops} key={index}>
                        {hops}
                      </option>
                    );
                  })}
                </select>
              </section>
              <section>
                <label htmlFor='malt' className='SearchForm__label'>
                  Malt:{' '}
                </label>
                <select
                  id='malt'
                  onChange={e => this.setState({ malt: e.target.value })}
                  value={this.state.malt ? this.state.malt : 'All'}
                  className='SearchForm__select'
                >
                  <option value=''>All</option>
                  {this.getMaltOptions().map((malt, index) => {
                    return (
                      <option value={malt} key={index}>
                        {malt}
                      </option>
                    );
                  })}
                </select>
              </section>
              <section>
                <label htmlFor='yeast' className='SearchForm__label'>
                  Yeast:{' '}
                </label>
                <select
                  id='yeast'
                  onChange={e => this.setState({ yeast: e.target.value })}
                  value={this.state.yeast ? this.state.yeast : 'All'}
                  className='SearchForm__select'
                >
                  <option value=''>All</option>
                  {this.getYeastOptions().map((yeast, index) => {
                    return (
                      <option value={yeast} key={index}>
                        {yeast}
                      </option>
                    );
                  })}
                </select>
              </section>
            </section>
            <button
              type='button'
              className='SearchForm__button--search'
              onClick={this.handleClick}
            >
              Search
            </button>
          </div>
        </form>
      </>
    );
  }
}

export const mapStateToProps = ({ beers, searchResults }) => ({
  beers,
  searchResults
});

export const mapDispatchToProps = dispatch => {
  return bindActionCreators({ setSearchResults }, dispatch);
};

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(SearchForm)
);

SearchForm.propTypes = {
  beers: PropTypes.array,
  searchResults: PropTypes.array,
  setSearchResults: PropTypes.func
};
