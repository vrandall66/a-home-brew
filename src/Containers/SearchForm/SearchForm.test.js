import React from 'react';
import { shallow } from 'enzyme';
import { setSearchResults } from '../../actions';
import { SearchForm, mapStateToProps, mapDispatchToProps } from './SearchForm';

describe('SearchForm', () => {
  let wrapper;
  let beers = [
    {
      ingredients: {
        hops: [
          { name: 'Amarillo' },
          { name: 'Fuggles' },
          { name: 'First Gold' }
        ],
        malt: [{ name: 'Brown' }, { name: 'Wheat' }, { name: 'Amber' }],
        yeast: 'yeast'
      }
    },
    {
      ingredients: {
        hops: [{ name: 'Amarillo' }, { name: 'Fuggles' }, { name: 'Simcoe' }],
        malt: [{ name: 'Brown' }, { name: 'Flaked Oats' }, { name: 'Amber' }],
        yeast: 'yeast'
      }
    },
    {
      ingredients: {
        hops: [
          { name: 'Amarillo' },
          { name: 'First Gold' },
          { name: 'Simcoe' }
        ],
        malt: [
          { name: 'Flaked Oats' },
          { name: 'Wheat' },
          { name: 'Caramalt' }
        ],
        yeast: 'yeast'
      }
    }
  ];
  const mockEvent = {
    target: {
      name: 'searchTerm',
      value: 'Stout'
    }
  };

  beforeEach(() => {
    wrapper = shallow(<SearchForm beers={beers} />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should update it's state property onChange of searchTerm input", () => {
    wrapper.instance().handleChange(mockEvent);

    expect(wrapper.state()).toEqual({
      searchTerm: 'Stout',
      hops: '',
      malt: '',
      yeast: ''
    });
  });

  it('should call handleClick on submit', () => {
    wrapper.instance().handleSubmit = jest.fn();
    wrapper.instance().forceUpdate();
    wrapper.find('.SearchForm').simulate('submit');

    expect(wrapper.instance().handleSubmit).toHaveBeenCalled();
  });

  it('should call handleClick within handleSubmit', () => {
    const event = { preventDefault: () => {} };
    jest.spyOn(event, 'preventDefault');
    wrapper.instance().handleClick = jest.fn();
    wrapper.instance().handleSubmit(event);

    expect(event.preventDefault).toBeCalled();
    expect(wrapper.instance().handleClick).toHaveBeenCalled();
  });

  it('should get all of the types of hops from the fetched beer data', () => {
    const expected = ['Amarillo', 'First Gold', 'Fuggles', 'Simcoe'];

    let sortedHops = wrapper.instance().getHopsOptions();
    expect(sortedHops).toEqual(expected);
  });

  // it('should setState onChange of select input', () => {
  //   wrapper.instance().setState = jest.fn();
  //   wrapper.find('#hops').simulate('mouseDown', { button: 0 });
  //   wrapper.find('#all-hops').simulate('mouseDown', {
  //     button: 0
  //   })

  //   expect(wrapper.instance().setState).toHaveBeenCalled();

  // });
});

describe('mapStateToProps', () => {
  it('should return an object with a beers array and searchResults array', () => {
    const mockStoreState = {
      beers: [{}, {}, {}],
      searchResults: [{}]
    };

    const expected = {
      beers: [{}, {}, {}],
      searchResults: [{}]
    };

    const mappedProps = mapStateToProps(mockStoreState);

    expect(mappedProps).toEqual(expected);
  });
});

describe('mapDispatchToProps', () => {
  it('calls dispatch with a setSearchResults, when handleClick is called', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = setSearchResults([{}, {}, {}]);

    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.setSearchResults([{}, {}, {}]);

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });
});
