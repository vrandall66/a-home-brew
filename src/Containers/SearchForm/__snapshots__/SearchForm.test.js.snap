// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm should match the snapshot 1`] = `
<Fragment>
  <form
    className="SearchForm"
    onSubmit={[Function]}
  >
    <input
      className="SearchForm__input--search-term"
      name="searchTerm"
      onChange={[Function]}
      placeholder="Search by name or description"
      value=""
    />
    <section>
      <label
        htmlFor="hops"
      >
        Hops:
      </label>
      <select
        classNamePrefix="list"
        id="hops"
        name="hops"
        onChange={[Function]}
        value="All"
      >
        <option
          id="all-hops"
          value=""
        >
          All
        </option>
        <option
          key="0"
          value="Amarillo"
        >
          Amarillo
        </option>
        <option
          key="1"
          value="First Gold"
        >
          First Gold
        </option>
        <option
          key="2"
          value="Fuggles"
        >
          Fuggles
        </option>
        <option
          key="3"
          value="Simcoe"
        >
          Simcoe
        </option>
      </select>
    </section>
    <section>
      <label
        htmlFor="malt"
      >
        Malt:
      </label>
      <select
        id="malt"
        onChange={[Function]}
        value="All"
      >
        <option
          value=""
        >
          All
        </option>
        <option
          key="0"
          value="Amber"
        >
          Amber
        </option>
        <option
          key="1"
          value="Brown"
        >
          Brown
        </option>
        <option
          key="2"
          value="Caramalt"
        >
          Caramalt
        </option>
        <option
          key="3"
          value="Flaked Oats"
        >
          Flaked Oats
        </option>
        <option
          key="4"
          value="Wheat"
        >
          Wheat
        </option>
      </select>
    </section>
    <section>
      <label
        htmlFor="yeast"
      >
        Yeast:
      </label>
      <select
        id="yeast"
        onChange={[Function]}
        value="All"
      >
        <option
          value=""
        >
          All
        </option>
        <option
          key="0"
          value="yeast"
        >
          yeast
        </option>
      </select>
    </section>
    <button
      onClick={[Function]}
      type="button"
    >
      Search
    </button>
  </form>
  <h5>
    You searched for: 
     
     
     
  </h5>
</Fragment>
`;
